src/qt/test/apptests.cpp

```diff
@@ -0,0 +1,92 @@
+#include "apptests.h"
+
+#include "chainparams.h"
+#include "init.h"
+#include "qt/bitcoin.h"
+#include "qt/bitcoingui.h"
+#include "qt/networkstyle.h"
+#include "qt/rpcconsole.h"
+#include "validation.h"
+
+#if defined(HAVE_CONFIG_H)
+#include "config/bitcoin-config.h"
+#endif
+#ifdef ENABLE_WALLET
+#include "wallet/db.h"
+#endif
+
+#include <QAction>
+#include <QEventLoop>
+#include <QLineEdit>
+#include <QScopedPointer>
+#include <QTest>
+#include <QTextEdit>
+#include <new>
+#include <string>
+#include <univalue.h>
+
+namespace {
+//! Call getblockchaininfo RPC and check first field of JSON output.
+void TestRpcCommand(RPCConsole* console)
+{
+    QLineEdit* lineEdit = console->findChild<QLineEdit*>("lineEdit");
+    lineEdit->setText("getblockchaininfo");
+    lineEdit->returnPressed();
+    QEventLoop loop;
+    QTextEdit* messagesWidget = console->findChild<QTextEdit*>("messagesWidget");
+    QObject::connect(messagesWidget, SIGNAL(textChanged()), &loop, SLOT(quit()));
+    loop.exec();
+    QString output = messagesWidget->toPlainText();
+    UniValue value;
+    value.read(output.right(output.size() - output.lastIndexOf(u8"\uFFFC") - 1).toStdString());
+    QCOMPARE(value["chain"].get_str(), std::string("regtest"));
+}
+}
+
+//! Entry point for BitcoinApplication tests.
+void AppTests::appTests()
+{
+    m_app.parameterSetup();
+    m_app.createOptionsModel(true /* reset settings */);
+    QScopedPointer<const NetworkStyle> style(
+        NetworkStyle::instantiate(QString::fromStdString(Params().NetworkIDString())));
+    m_app.createWindow(style.data());
+    BitcoinCore::baseInitialize();
+    expectCallback(); // windowShown
+    connect(&m_app, SIGNAL(windowShown(BitcoinGUI*)), this, SLOT(guiTests(BitcoinGUI*)));
+    m_app.requestInitialize();
+    m_app.exec();
+
+    // Reset some global state to avoid interfering with later tests.
+    ResetShutdownRequested();
+#ifdef ENABLE_WALLET
+    ::bitdb.~CDBEnv();
+    new (&::bitdb) CDBEnv;
+#endif
+    UnloadBlockIndex();
+}
+
+//! Entry point for BitcoinGUI tests.
+void AppTests::guiTests(BitcoinGUI* window)
+{
+    HandleCallback windowShown{*this};
+    expectCallback(); // consoleShown
```

Comment:
How about storing the name? Is this supposed to be stack like?

Reply:
> How about storing the name? Is this supposed to be stack like?

Kind of, but it's a pool, not a stack. The idea is just for AppTests to just create a bunch of windows and let appropriate tests run as the windows are created, without having to worry about the order events happen. To track names instead of using a counter, `int m_callbacks` could be replaced with `std::multiset<std::string> m_callbacks`, and `++` `--` replaced by `insert` and `erase`. It seems like this just be more verbose and error prone without being more readable the current comments, but I could change it you think it would be better.

Reply:
I see 2 advantages over counting:
 - the comment is optional but the argument is not;
 - Better hint if the set is not empty at the end.

BTW, have you thought using QSignalSpy?

Reply:
> I see 2 advantages over counting:
>
> * the comment is optional but the argument is not;
> * better hint if the set is not empty at the end.
>
> BTW, have you thought using QSignalSpy?

Ok, switched from an int to a set. I didn't know about signal spy, but it doesn't seem quite right for this because it can track signals but not run handlers. Here I am tracking signals (to trigger shutdown after last one) but only in the course of handling them. Signal spy does have a wait method which looks like it would be nice to replace the temporary loop in TestRpcCommand, but unfortunately it is in qt5 but not qt4.